src/sdk/namespace_fs.js

```diff
@@ -492,23 +494,94 @@ class NamespaceFS {
     ///////////////////
 
     async upload_object(params, object_sdk) {
-        try {
-            const fs_account_config = this.set_cur_fs_account_config(object_sdk);
-            await this._load_bucket(params, fs_account_config);
+        const fs_account_config = this.set_cur_fs_account_config(object_sdk);
+        await this._load_bucket(params, fs_account_config);
 
-            const file_path = this._get_file_path(params);
-            const upload_id = uuidv4();
-            const upload_path = path.join(this.bucket_path, config.NSFS_TEMP_DIR_NAME, 'uploads', upload_id);
-            // console.log('NamespaceFS.upload_object:', upload_path, '->', file_path);
+        const file_path = this._get_file_path(params);
+        // Uploading to a temp file then we will rename it. 
+        const upload_id = uuidv4();
+        const upload_path = path.join(this.bucket_path, config.NSFS_TEMP_DIR_NAME, 'uploads', upload_id);
+        // console.log('NamespaceFS.upload_object:', upload_path, '->', file_path);
+        await Promise.all([this._make_path_dirs(file_path, fs_account_config), this._make_path_dirs(upload_path, fs_account_config)]);
+        if (params.copy_source) {
+            const source_file_path = path.join(this.bucket_path, params.copy_source.key);
+            try {
+                const same_inode = await this._is_same_inode(fs_account_config, source_file_path, file_path);
+                if (same_inode) return same_inode;
+                // Doing a hard link.
+                await nb_native().fs.link(fs_account_config, source_file_path, upload_path);
+            } catch (e) {
+                try {
+                    console.warn('NamespaceFS: COPY using link failed with:', e);
+                    await this._copy_stream(source_file_path, upload_path, fs_account_config);
+                } catch (err) {
+                    throw this._translate_object_error_codes(err);
```

Comment:
This catch is the last catch for both IF cases, you can handle both IF blocks with this single catch instead of multiplying it and creating more nested try/catch blocks.